crates/sui-core/src/authority/authority_notifier.rs

```diff
@@ -51,16 +51,17 @@ impl TransactionNotifier {
         self.low_watermark.load(Ordering::SeqCst)
     }
 
-    /// Check that we have drained all tickets (i.e. low watermark reached high watermark),
-    /// and that the issued tickets ends before `end`.
-    pub fn ticket_drained_til(&self, end: TxSequenceNumber) -> bool {
+    /// Check that we have drained all tickets (i.e. low watermark reached high watermark).
+    /// Return the watermark if we have drained.
+    pub fn ticket_drained(&self) -> Option<u64> {
         let high_watermark = self.inner.lock().high_watermark;
         let low_watermark = self.low_watermark.load(Ordering::SeqCst);
```

Comment:
Are we sure this is not a race condition? We load two values without holding a lock common to both.

I'm guessing that the answer is "the validator is halted so high_watermark cannot advance" - but is that actually true? We are still processing transactions trying to close out the epoch, right?

Reply:
At the time this function is called (from reconfiguration), the validator is already halted, which means `high_watermark` will not move at while this function is being called. So it should be safe.
The `high_watermark` can advance latter once we start syncing the missing transactions in the last checkpoint, but that won't happen until we go pass this point.

Reply:
I believe you, but this seems like a very fragile invariant in that it fairly innocent changes in the future could break it. Can we instead put the loop inside this function, so that it can hold the high_watermark lock the entire time it is waiting?

Reply:
hmm I am not sure how to do that.
There will be a async sleep in the loop, and you cannot hold a lock to the high_watermark while async wait.

Reply:
you can if it is a tokio mutex

Reply:
There is an easier solution here though: I just need to hold the lock while checking the `low_watermark`, the loop can stay outside safely.

Reply:
Ok - approving now, I think one of those solutions would be a good idea but I don't want to hold this up any longer.